File: backend/src/main/java/application/entities/Order.java
package application.entities;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import static jakarta.persistence.FetchType.EAGER;
import java.util.*;

@Getter
@Setter
@EqualsAndHashCode
@Entity
@Table(name = "\"order\"")
public class Order {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    private Long id;

    @ManyToOne(fetch = EAGER)
    @JoinColumn(name = "user_id")
    @JsonBackReference(value = "user-orders")
    private Utente user;

    @Basic
    @NotNull
    @Column(name = "total")
    private Double total;

    @Basic
    @Temporal(TemporalType.TIMESTAMP)
    @CreationTimestamp
    @Column(name = "create_date")
    private Date createDate;

    @Basic
    @Column(name = "status")
    private String status="Pending";

    @ToString.Exclude
    @OneToMany(fetch = EAGER,mappedBy = "order",cascade = {CascadeType.REMOVE,CascadeType.MERGE})
    @JsonManagedReference(value = "orderProducts")
    private List<OrderProducts> orderProducts;

    @JsonBackReference(value = "order")
    @OneToOne(mappedBy = "order")
    private Cart cart;


}

File: frontend/src/app/models/Order.ts
import { OrderProducts } from "./Orderproducts";
import { User, UserImpl } from "./User";
export interface OrderDTO {
  id: number;
  userId: User;
  total: number;
  createDate: Date;
  status: string;
  orderProducts: OrderProducts[];
}

export interface Order {
  id: number;
  user: User;
  total: number;
  createDate: Date;
  status: string;
  orderProducts: OrderProducts[];
}

export class OrderImpl implements Order{
  public id: number;
  public user: User;
  public total: number;
  public createDate: Date;
  public status: string;
  public orderProducts: OrderProducts[];

  constructor(){
    this.id=1;
    this.user=new UserImpl();
    this.total=0.0;
    this.createDate=new Date();
    this.status='Pending';
    this.orderProducts=[];
  }

  copy(c:Order) : OrderImpl{
    this.id=c.id;
    this.user=c.user;
    this.total=c.total;
    this.createDate=c.createDate;
    this.status=c.status;
    this.orderProducts=c.orderProducts;
    return this;
  }

  copyDTO(c:OrderDTO) : OrderImpl{
    this.id=c.id;
    this.user=c.userId;
    this.total=c.total;
    this.createDate=c.createDate;
    this.status=c.status;
    this.orderProducts=c.orderProducts;
    return this;
  }
}

export class OrderDTOImpl implements OrderDTO{
  public id: number;
  public userId: User;
  public total: number;
  public createDate: Date;
  public status: string;
  public orderProducts: OrderProducts[];

  constructor(){
    this.id=1;
    this.userId=new UserImpl();
    this.total=0.0;
    this.createDate=new Date();
    this.status='Pending';
    this.orderProducts=[];
  }

  copy(c:OrderDTO) : OrderDTOImpl{
    this.id=c.id;
    this.userId=c.userId;
    this.total=c.total;
    this.createDate=c.createDate;
    this.status=c.status;
    this.orderProducts=c.orderProducts;
    return this;
  }

  copyOrder(c:Order) : OrderDTOImpl{
    this.id=c.id;
    this.userId=c.user;
    this.total=c.total;
    this.createDate=c.createDate;
    this.status=c.status;
    this.orderProducts=c.orderProducts;
    return this;
  }

}
